src/browser/speaker.js

```diff
@@ -19,145 +22,1121 @@ function SpeakerAdapter(bus)
         return;
     }
 
-    /** @const @type {BusConnector} */
+    var SpeakerDAC;
+    if(window.AudioWorklet)
+    {
+        SpeakerDAC = SpeakerWorkletDAC;
+    }
+    else
+    {
+        SpeakerDAC = SpeakerBufferSourceDAC;
+    }
+
+    /** @const */
     this.bus = bus;
 
+    /** @const */
     this.audio_context = new (window.AudioContext || window.webkitAudioContext)();
 
-    this.beep_gain = this.audio_context.createGain();
-    this.beep_gain.gain.setValueAtTime(0, this.audio_context.currentTime);
-    this.beep_gain.connect(this.audio_context.destination);
-
-    this.beep_oscillator = this.audio_context.createOscillator();
-    this.beep_oscillator.type = "square";
-    this.beep_oscillator.frequency.setValueAtTime(440, this.audio_context.currentTime);
-    this.beep_oscillator.connect(this.beep_gain);
-    this.beep_oscillator.start();
-
-    this.beep_playing = false;
-    this.beep_enable = false;
-    this.beep_frequency = 440;
-    this.pit_enabled = false;
-    this.cpu_running = true;
-
-    this.dac_processor = this.audio_context.createScriptProcessor(2048, 0, 2);
-    this.dac_processor.onaudioprocess = this.dac_process.bind(this);
-    this.dac_processor.connect(this.audio_context.destination);
-    this.dac_buffer0 = new Float32Array(this.dac_processor.bufferSize);
-    this.dac_buffer1 = new Float32Array(this.dac_processor.bufferSize);
-    this.dac_enabled = true;
+    /** @const */
+    this.mixer = new SpeakerMixer(bus, this.audio_context);
+
+    /** @const */
+    this.pcspeaker = new PCSpeaker(bus, this.audio_context, this.mixer);
+
+    /** @const */
+    this.dac = new SpeakerDAC(bus, this.audio_context, this.mixer);
 
     bus.register("emulator-stopped", function()
     {
-        this.cpu_running = false;
-        this.beep_update();
+        this.audio_context.suspend();
     }, this);
 
     bus.register("emulator-started", function()
     {
-        this.cpu_running = true;
-        this.beep_update();
+        this.audio_context.resume();
+    }, this);
+
+}
+
+/**
+ * @constructor
+ * @param {!BusConnector} bus
+ * @param {!AudioContext} audio_context
+ */
+function SpeakerMixer(bus, audio_context)
+{
+    /** @const */
+    this.audio_context = audio_context;
+
+    this.sources = new Map();
```

Comment:
I think so. I like integer constants too.

The strings aren't used elsewhere apart from some constructors and the `mixer-*` bus messages to identify which mixer source to modify.

Since `browser/speaker.js` is only an implementation that handles the appropriate bus messages (i.e. there could potentially be a `node/speaker.js` that handles the same bus messages), I think we will need to define the integer constants in some place outside of `browser/speaker.js`, in some place visible to `src/sb16.js` for example.

Would [`src/const.js`](https://github.com/copy/v86/blob/master/src/const.js) be the right place for this?

Reply:
> Would src/const.js be the right place for this?

Yeah.